server/service/src/sensor/update.rs

```diff
@@ -54,10 +54,14 @@ pub fn validate(
     }
 
     if let Some(location_id) = &input.location_id {
-        match check_location_on_hold(&location_id, connection) {
-            Ok(true) => return Err(UpdateSensorError::LocationIsOnHold),
-            Err(e) => return Err(UpdateSensorError::DatabaseError(e)),
-            _ => (),
+        // Check if location on hold only if location is not "none" value
+        // None passed from client is the unassignment of location from sensor node
+        if location_id != &"None".to_string() {
```

Comment:
Hmm, are we actually sending 'None' value through to back end ? I think 'None' value should be an option in the drop down, but we should have 'unlink location from sensor' mutation

Reply:
Or, we would need to implement optional not nullable input which is tricky in graphql input, but it would be easier if we apply this to other areas

https://github.com/openmsupply/remote-server/issues/671

Reply:
Think should allow passing null to location_id since we want to implement this in a few places maybe? 🤔 None of the location usage right now has a 'remove from location' option

Reply:
I'm considering two different options:

- 1 Add the nested optional value as per https://github.com/openmsupply/remote-server/issues/671 . Possibly a little complex?
or 
- 2 make the location_id always overwrite / a compulsory value, so that if it isn't supplied, it will overwrite to null

I'm not sure what is best given the context that we might be using this all places where a location can be assigned.

I wonder if @andreievg or @roxy-dao have any thoughts?

@clemens-msupply FYI :)

Reply:
I would go for one. 

I wouldn't make location_id a compulsory value. Something may have a location already saved and if you overwrite this with null, then the location will be gone....

Reply:
Yes agree, yesterday I found it quite complicated to explain and wondered if we just go for a simpler solution, but this would create an API that is easier to missuses and would result in data loose...

Think we should use something like in the linked issue. I added some docs written to be read as an API user, i.e. I justed TypeScript syntax:

```rust
/// Update a nullable value
///
/// This struct is usually used as an optional value.
/// For example, in an API update input object like `mutableValue:  NullableUpdate | null | undefined`.
/// This is done to encode the following cases (using `mutableValue` from previous example):
/// 1) if `mutableValue` is `null | undefined`, nothing is updated
/// 2) if `mutableValue` object is set:
///   a) if `NullableUpdate.value` is `undefined | null`, the `mutableValue` is set to `null`
///   b) if `NullableUpdate.value` is set, the `mutableValue` is set to the provided `NullableUpdate.value`
#[derive(InputObject)]
pub struct NullableUpdate<T> {
    pub value: Option<T>,
}
```

Reply:
Yeah i like the generic solution too, I think I had issues with it though, when used across crates (maybe it was in earlier versions of async-graqphl, the generic structure is also easy to [apply to other types](https://async-graphql.github.io/async-graphql/en/define_simple_object.html?highlight=graphql(concrete(#generic-simpleobjects).

Anyways if it's possible would prefer a generic structure with concrete, otherwise still a generic structure but with types

`pub struct NullableStringUpdate`...

Reply:
@fergie-nz can have a quick chat about this

Reply:
Sounds good!

I've just added it as a generic LocationInput (with the same structure as the optional and nullable TaxInput). I could make this generic to any other nullable strings.

I'm also wondering how to approach (or if I need to approach) mSupply end points. I've made all places which were previously supplied a location_id now accept a nullable location. I'm wondering if some of these are mSupply end points?  Or perhaps they're all for omSupply but yet unused in the front end?